File: app/src/main/java/com/github/ivanshafran/unit_tests_evolution/BuyActivity.java
package com.github.ivanshafran.unit_tests_evolution;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.View;

public class BuyActivity extends AppCompatActivity implements RateUsDialog.Listener {

    private BuyPreferences buyPreferences;
    private ShowRateUsLogic showRateUsLogic;

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_buy);

        buyPreferences = new BuyPreferences(this);
        showRateUsLogic = new ShowRateUsLogic(
                new RateUsPreferences(this),
                buyPreferences,
                new Time()
        );

        findViewById(R.id.buyButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {
                buyPreferences.incrementBuyCount();
                if (showRateUsLogic.shouldShowRateUs()) {
                    showRateUs();
                }
            }
        });
    }

    private void showRateUs() {
        new RateUsDialog().show(getSupportFragmentManager(), null);
    }

    @Override
    public void onRateNowClick() {
        // Go to Google Play
    }
}

File: app/src/main/java/com/github/ivanshafran/unit_tests_evolution/BuyActivity.kt
package com.github.ivanshafran.unit_tests_evolution

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View

class BuyActivity : AppCompatActivity(), RateUsDialog.Listener {

    private lateinit var buyPreferences: BuyPreferences
    private lateinit var showRateUsLogic: ShowRateUsLogic

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_buy)

        buyPreferences = BuyPreferences(this)
        showRateUsLogic = ShowRateUsLogic(
            RateUsPreferences(this),
            buyPreferences,
            Time()
        )

        findViewById<View>(R.id.buyButton).setOnClickListener {
            buyPreferences.incrementBuyCount()
            if (showRateUsLogic.shouldShowRateUs()) {
                showRateUs()
            }
        }
    }

    private fun showRateUs() {
        RateUsDialog().show(supportFragmentManager, null)
    }

    override fun onRateNowClick() {
        // Go to Google Play
    }
}
